main.cpp

```diff
@@ -17,41 +17,79 @@ Purpose:  This project will teach you about variadic templates and recursive tem
 Make the following program work, which makes use of Variadic templates and Recursion
  */
 
-#include <iostream>
-#include <string>
-#include <typeinfo>
-
-struct Point
-{
-    Point(float _x, float _y) : x(_x), y(_y) { }
-    Point& multiply(float m)
-    {
-        x *= m;
-        y *= m;
-        return *this;
-    }
-    std::string toString() const
-    {
-        std::string str;
-        str +="Point { x: ";
-        str += std::to_string(x);
-        str += ", y: ";
-        str += std::to_string(y);
-        str += " }";
-        return str;
-    }
-private:
-    float x{0}, y{0};
-};
-
-template<typename Type>
-struct Wrapper
-{
-    Wrapper(Type&& t) : val(std::move(t)) 
-    { 
-        std::cout << "Wrapper(" << typeid(val).name() << ")" << std::endl; 
-    }
-};
+#include <iostream>
+#include <string>
+#include <typeinfo>
+#include "typename.h"
+
+struct Point
+{
+    Point(float _x, float _y) : x(_x), y(_y) { }
+    Point& multiply(float m)
+    {
+        x *= m;
+        y *= m;
+        return *this;
+    }
+    std::string toString() const
+    {
+        std::string str;
+        str +="Point { x: ";
+        str += std::to_string(x);
+        str += ", y: ";
+        str += std::to_string(y);
+        str += " }";
+        return str;
+    }
+private:
+    float x{0}, y{0};
+};
+
+template<typename Type>
+struct Wrapper
+{
+    Wrapper(Type&& t) : val(std::move(t))
+    { 
+        std::cout << "Wrapper(" << type_name<decltype(val)>() << ")" << std::endl;
+    }
+
+    void print()
+    {
+        std::cout << "Wrapper::print(" << val << ")" << std::endl;
+    }
+
+private:
+    Type val;
+};
+
+template<>
+struct Wrapper<Point>
+{
+    Wrapper(Point&& p) : point(std::move(p))
+    { 
+        std::cout << "Wrapper(" << type_name<decltype(point)>() << ")" << std::endl;
+    }
+
+    void print()
+    {
+        std::cout << "Wrapper::print(" << point.toString() << ")" << std::endl;
+    }
+private:
+    Point point;
+};
+
+template<typename T, typename ... Args>
+void variadicHelper(T&& first, Args&& ... args)
+{
+    Wrapper<T>(std::forward<T>(first)).print();
+    variadicHelper( std::forward<Args>(args) ... );
+}
+
+template<typename T>
+void variadicHelper(T&& first)
+{
+    Wrapper<T>(std::forward<T>(first)).print();
+}

```

Comment:
Project9.1 challenge 2:
There is another recursive solution.  it is possible for `Args ...` to be empty.
your task: remove your implementation for instruction `4)` and implement this alternate solution.
implement it **after** your existing `variadicHelper` function.
hint: what does the **recursive call** actually look like when called with an empty Variadic Parameter list?